### PR TITLE
Fix nil KVOChange oldValue/newValue for RawRepresentable types

### PR DESCRIPTION
**Pull request by ffried**
This fixes the `oldValue` / `newValue` properties of `NSKeyValueObservedChange<Value>` being always `nil` when key-value-observing `RawRepresentable` types (typically enums) like `URLSessionTask.State`.

Resolves [SR-5872](https://bugs.swift.org/browse/SR-5872).

### CODE REVIEW

**Review by Catfish-Man — APPROVED** [review]
The code duplication is rather unfortunate, but I can't think of a better alternative either. Thank you for fixing this!

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test macOS platform

**Comment by Catfish-Man** [comment]
@swift-ci please test OS X platform

**Comment by ffried** [comment]
Rebased because of conflicts with `master`.

/cc @millenomi

**Comment by jrose-apple** [comment]
(just kicking off tests, not reviewing)

@swift-ci Please test

**Comment by ffried** [comment]
Rebased again to fix conflicts with `master`...

Is there anything I can do to get this merged?

/cc @millenomi @jrose-apple

**Review by millenomi — COMMENTED** [review]
I’m looking, but this adds API surface, which is generally an issue.

**Comment by jrose-apple** [comment]
I am a little concerned about using the static type of the key path to decide this rather than the dynamic type.

**Comment by ffried** [comment]
@jrose-apple I've tried to come up with an example where this would be an issue, but I couldn't find one. The closest I could find was when the `Value` of the key path is some (`@objc`) protocol existential where the concrete type is in fact `RawRepresentable`. But I suppose this happens less often than we have unexpected `nil` values now.

Is there even a way to get the raw value type of a dynamic type? Since `RawRepresentable` has `associatedtype` constraints, we can't really get to its `RawValue` from a dynamic type, or am I missing something?

**Comment by jrose-apple** [comment]
I don't have a solution for you, but it's an issue as soon as you wrap the observing function in another layer.

```swift
extension NSObject {
  func observeByPrinting<Value>(
    _ keyPath: KeyPath<Self, Value>
  ) -> NSKeyValueObservation {
    return self.observe(keyPath, options: [.new]) {
      print("CHANGED \($1.newValue!)")
    }
  }
}
```

(Generics aren't like C++ templates; there's one implementation at runtime, and so it has to use the most general overload.)

Maybe it's better to take this to fix the common case and then not close the bug until we get generalized existentials.

**Comment by ffried** [comment]
@jrose-apple I see, thanks for the example.
The difference here is though, that your case is fixable by whoever wrote the code (by just adding the necessary overloads). But the problem I'm addressing in this PR can't be fixed in "external" code.
Having the stdlib cover all cases is nicer, of course.
But just out of curiosity: Aren't overloads better for performance than dynamic type checks?

IMHO we should close the bug and open a new one for the overload case. I'm not sure how big of a problem it will be anyway (nor if this is really a "bug" in Swift rather). I think for most people this fix will suffice.
But I'd also be fine with leaving the bug open...

**Comment by jrose-apple** [comment]
That's a reasonable point.

Overloads are better than dynamic checks in non-inlinable code, sure, but they also have a code size cost. Even if that code size is just part of the operating system going forward.

**Comment by ffried** [comment]
Any news on this?

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by ffried** [comment]
@millenomi Tests are green. 🙂
Anything left to do before this can be merged?

**Comment by ffried** [comment]
@jrose-apple One thought about leaving the bug report open came to my mind today:
How are the Swift release notes created? I think we should mention this fix (even if it's not "final" yet), since I think that folks have started to build their own workarounds to this problems (using wrapper objects or anything the like).

**Comment by jrose-apple** [comment]
You can add things to CHANGELOG.md. Seems reasonable for this case.

(The Apple-side release notes go through a larger process, but CHANGELOG is a source of information for them.)

**Comment by ffried** [comment]
@jrose-apple Ah, good to know, thanks!
Should I add this to the Swift 5.1 section (and create a PR against the swift-5.1-branch)?
Or what's the correct version / headline to add it to?

**Comment by jrose-apple** [comment]
I think that makes sense, yeah. This is definitely worth getting into 5.1 (heck, we're sorry it missed 5.0!), so let's be optimistic in the CHANGELOG. :-)

**Comment by ffried** [comment]
Ok, I'll quickly rebase and update the CHANGELOG.md then. Thanks!

**Comment by ffried** [comment]
@jrose-apple: Done. Branch is rebased and CHANGELOG is updated. Could you give @swift-ci a nudge?

And #23794 is the PR against the Swift 5.1 branch.

**Comment by jrose-apple** [comment]
We usually wait until the master PR lands before doing anything with the cherry-pick. @millenomi does still have to sign off on this and she's been swamped but will hopefully claw her way back to the surface soon.

@swift-ci Please smoke test

**Comment by ffried** [comment]
Oh, I was a little bit too quick with the cherry pick then - sorry. Is it fine to leave it just laying around or should I close it again and wait until this is merged?

**Comment by jrose-apple** [comment]
I guess it'd be polite to close so that the release managers don't wonder what's going on with it. You could leave the branch there so that it's easy to re-open, though.

**Comment by ffried** [comment]
Alright, thanks! I’ve closed it again.

**Comment by millenomi** [comment]
This patch increases API surface, and it needs to undergo internal review first. Sorry for the delay!

**Comment by parkera** [comment]
Just to add on to that: it shouldn't be a problem, we just want to get some additional eyes on it to make sure everything is correct.

**Comment by ffried** [comment]
_Rebased because of conflicts._

**Comment by ffried** [comment]
@millenomi @parkera: Quick question: Should I keep pinging you about this? Or is this anyway on some internal list and I'd only be generating unnecessary noise?

**Comment by parkera** [comment]
The reminders still help. :)

**Comment by nh7a** [comment]
I was trying to observe `AVPlayerItem.status` and hit this bug. Ouch :-p

**Comment by Jakartoine** [comment]
Same as @nh7a , I hit this bug too.

**Comment by ffried** [comment]
@parkera @millenomi Ping 🙂

**Comment by ffried** [comment]
@millenomi @parkera Ping... 🙃

**Comment by jrose-apple** [comment]
Eavesdrop testing.

@swift-ci Please smoke test

**Comment by s678g10** [comment]
gggg

**Comment by ffried** [comment]
@millenomi @parkera Ping...

I just wondered, do we have a similar problem with the new `KeyValueObservingPublisher`?

**Comment by nh7a** [comment]
Any update on this? I really hope this to get into 5.1.

**Comment by aabdellah** [comment]
> Any update on this? I really hope this to get into 5.1.

I think it's too late for 5.1 now.

**Comment by ffried** [comment]
@jrose-apple @parkera @millenomi @Catfish-Man 
Since my last rebase, a fix was made for SR-6066 that made sure that `oldValue`/`newValue` contain `.some(nil)` for optional types.

This broke my initial version of this fix, since for optional RawRepresentable types, the cast would still be attempted to the type itself instead of its RawValue.
Further, I've discovered that in my overload for optional RawRepresentable types, the optionals were accidentally flattened as well.
The fix for SR-6066 was basically trying to cast `NSNull` to the observed type (which of course always fails) and was then using that `nil` value in another cast to the observed type, which succeeds when the observed type is also an optional.
I've simplified that by removing the protocol and directly attempting the cast from `Optional<Any>.none` to `Value` in that case.
This also fixes a bug introduced in the previous version, where `oldValue` would be `.some(nil)` instead of `nil` for the `.initial` callback (which should only have `newValue` being not nil, in case the `.new` option was given).

Long story short, the basic concept of my initial fix is still the same (having the overloads and the `valueConverter` closure), but the implementation now touches a little bit more code than before. Thus I'd appreciate it, if @Catfish-Man and @millenomi could have another look at that.

@parkera @jrose-apple Is there still any hope for this making it into Swift 5.1?
Also, do you guys know about `NSObject.KeyValueObservingPublisher`? I only see one overload there as well. Did whoever implemented it solve the RawRepresentable problem in a different/better way? Or does it have this problem as well?

**Comment by CodaFi** [comment]
This could still make 5.1.1, but it's going to be tight.  If you're still interested in this change, please rebase this pull request and I'll make sure that the relevant parties are CC'd here.

**Comment by ffried** [comment]
@CodaFi Thanks for the heads-up. I've just rebased this PR. By "the Swift after 5.1.1" do you mean Swift 5.2? Or the next 5.1.x? I'm asking because of the CHANGELOG entry, which is still in the Swift 5.1 section now. Should I move that to the Swift 5.2 section?

**Comment by CodaFi** [comment]
Tagging @parkera because this requires Foundation’s stamp to ship.

**Comment by nh7a** [comment]
@parkera any update on this? Is it possible this PR to get merged by Swift 6?

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by strawb3rryx7** [comment]
Any updates on this?

**Comment by ffried** [comment]
@parkera @CodaFi @millenomi Any news on this?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by nstefan** [comment]
Hum this issue leads to very odd behaviours indeed… I've been dragged here while investigaing on AVFoundation.
Thank you for submiting a fix 🙏. I hope this gets more attention.

**Comment by ffried** [comment]
The overlay apparently moved to swift-corelibs-foundation. I've opened a PR there, which might even fix this issue w/o adding any new public API.

@parkera @millenomi @Catfish-Man Can one of you guys please have a look and maybe help me with the testing issues there?

**Comment by BudhirajaRajesh** [comment]
Any update on this?
